src/csg_tree.cpp

```diff
@@ -565,77 +444,239 @@ void CsgOpNode::BatchUnion() const {
+  // differ in `transform_`, so we want it to be able to share the result.
+  while (!stack.empty()) {
+    std::shared_ptr<CsgStackFrame> frame = stack.back();
+    // Because `CsgOpNode` may be shared and we may encounter some evaluated
+    // nodes that are evaluated during the DFS, we can skip calculation by
+    // checking for `cache_` early.
+    // This is probably not a very useful optimization, because in those cases
+    // the `children_` set only contains one element, and `BatchUnion`,
+    // `BatchBoolean` or `Subtract` on this is already a no-op...
+    if (frame->op_node->cache_) {
+      frame->destination->push_back(std::static_pointer_cast<CsgLeafNode>(
+          frame->op_node->cache_->Transform(frame->transform)));
+      stack.pop_back();
+      continue;
+    }
+    auto impl = frame->op_node->impl_.GetGuard();
+    if (frame->finalize) {
+      if (frame->op_node->cache_ == nullptr) switch (frame->op_node->op_) {
+          case OpType::Add:
+            BatchUnion(frame->left_children);
+            impl->children_ = {frame->left_children[0]};
+            break;
+          case OpType::Intersect: {
+            std::vector<std::shared_ptr<const Manifold::Impl>> impls;
+            for (auto &child : frame->left_children)
+              impls.push_back(child->GetImpl());
+            auto result = BatchBoolean(OpType::Intersect, impls);
+            impl->children_ = {std::make_shared<CsgLeafNode>(result)};
+            break;
+          };
+          case OpType::Subtract:
+            if (frame->left_children.empty()) {
+              // nothing to subtract from, so the result is empty.
+              impl->children_ = {std::make_shared<CsgLeafNode>()};
+            } else {
+              BatchUnion(frame->left_children);
+              if (frame->right_children.empty()) {
+                // nothing to subtract, result equal to the LHS.
+                impl->children_ = {frame->left_children[0]};
+              } else {
+                BatchUnion(frame->right_children);
+                Boolean3 boolean(*frame->left_children[0]->GetImpl(),
+                                 *frame->right_children[0]->GetImpl(),
+                                 OpType::Subtract);
+                Manifold::Impl result = boolean.Result(OpType::Subtract);
+                impl->children_ = {std::make_shared<CsgLeafNode>(
+                    std::make_shared<Manifold::Impl>(result))};
+              }
+            }
+            break;
+        }
+      frame->op_node->cache_ = std::static_pointer_cast<CsgLeafNode>(
+          impl->children_[0]->Transform(frame->op_node->transform_));
+      if (frame->destination != nullptr)
+        frame->destination->push_back(std::static_pointer_cast<CsgLeafNode>(
+            frame->op_node->cache_->Transform(frame->transform)));
+      stack.pop_back();
+    } else {
+      auto add_children = [&](std::shared_ptr<CsgNode> &node, OpType op,
```

Comment:
I prefer to be explicit about what's captured in a lambda - it helps me to understand the scope of the function.